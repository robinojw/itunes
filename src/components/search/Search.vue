<template>
  <div class="search-bar">
    <input
      type="text"
      class="search-bar"
      placeholder="Search"
      @focus="toggleInput"
      @blur="toggleInput"
      ref="input"
      v-model="inputValue"
    />
    <div class="helper-text" ref="helper">
      {{ inputValue }} search media with the format media/artist i.e album/oasis
      <p class="enter-text">press enter...</p>
    </div>
  </div>
</template>

<script lang="ts">
import './search.scss';
import { Component, Vue } from 'vue-property-decorator';

const HELPER_TEXT = 'show-helper';
const ENTER_KEY_CODE = 13;

@Component
export default class Search extends Vue {
  $refs!: {
    helper: HTMLElement;
  };
  focused = false;
  inputValue: string = window.location.pathname
    .slice(1)
    .replace(/[^a-zA-Z/]/g, ' ');

  toggleInput() {
    this.$refs.helper.classList.toggle(HELPER_TEXT);
    this.focused = !this.focused;
    this.toggleKeyboardListener();
  }

  toggleKeyboardListener() {
    if (this.focused) {
      window.addEventListener('keydown', (event: KeyboardEvent) =>
        this.searchMedia(event)
      );
    }
  }

  /**
   * Get the latest search parameters and emit an event to update the
   * parent's resource data.
   */
  searchMedia(event: KeyboardEvent) {
    if (
      event.keyCode == ENTER_KEY_CODE &&
      window.location.pathname != `/${this.inputValue}`
    ) {
      const [media, query] = this.inputValue.split('/');
      const value = this.inputValue.replace(/[^a-zA-Z/]/g, '+');
      this.$router.push(`/${value}`);
      this.$emit('update-resource', media, query);
      window.removeEventListener('keydown', this.searchMedia);
    }
  }
}
</script>
