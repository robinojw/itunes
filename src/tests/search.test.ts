import 'jest';
require('jsdom-global')();
import { mount } from '@vue/test-utils';
import Search from '../components/search/Search.vue';

const UPDATE_EVENT = 'update-resource';

describe('Mounted Search', () => {
  const wrapper = mount(Search);

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it('renders search component correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('emits a update-resource event', async () => {
    wrapper.vm.$emit(UPDATE_EVENT);
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted(UPDATE_EVENT)).toBeTruthy();
  });
});
