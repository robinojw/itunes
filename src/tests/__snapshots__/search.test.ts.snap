// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mounted App renders search component correctly 1`] = `
<div
  class="search-bar"
>
  <input
    class="search-bar"
    placeholder="Search"
    type="text"
  />
   
  <div
    class="helper-text"
  >
    
    blank search media with the format media/artist i.e album/oasis
    
    <p
      class="enter-text"
    >
      press enter...
    </p>
  </div>
</div>
`;
